tokenizers/src/models/unigram/model.rs

```diff
@@ -0,0 +1,433 @@
+use crate::models::unigram::lattice::Lattice;
+use crate::models::unigram::trie::{Trie, TrieBuilder};
+use crate::tokenizer::{Model, Result, Token};
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+
+type TokenMap = HashMap<String, u32>;
+type Vocab = Vec<String>;
+
+/// A `Unigram` model to encode sentences.
+#[derive(Clone)]
+pub struct Unigram {
+    token_to_ids: TokenMap,
+    pub(crate) vocab: Vocab,
+    pub(super) scores: Vec<f64>,
+    trie: Trie<char>,
+    pub min_score: f64,
+    pub(super) unk_id: usize,
+    pub(super) bos_id: usize,
+    pub(super) eos_id: usize,
+
+    fuse_unk: bool,
+}
+impl PartialEq for Unigram {
+    fn eq(&self, other: &Self) -> bool {
+        let vocab: Vec<(&String, &f64)> = self.vocab.iter().zip(self.scores.iter()).collect();
+        let other_vocab: Vec<(&String, &f64)> =
+            other.vocab.iter().zip(other.scores.iter()).collect();
+        self.unk_id == other.unk_id && vocab == other_vocab
+    }
+}
+
+impl std::fmt::Debug for Unigram {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct("BPE")
+            .field("vocab", &self.vocab.len())
+            .finish()
+    }
+}
+
+static K_UNK_PENALTY: f64 = 10.0;
+
+impl Default for Unigram {
+    fn default() -> Self {
+        let vocab = vec![("<unk>".to_string(), 0.0)];
+        Self::from(&vocab, 0)
+    }
+}
+
+impl Unigram {
+    /// Create a `Unigram` model from a given vocabulary.
+    /// Vocabulary are the various tokens and their associated score which is a sort of a logprob of
+    /// their frequency, which will enable tokenization and sampling.
+    /// unk_id, is the index within the vocabulary.
+    /// For now `Unigram` *requires* at least `unk` because we might find a never seen char.
+    /// Further versions might allow that part to be hidden.
+    pub fn from(vocabulary: &[(String, f64)], unk_id: usize) -> Self {
+        let n = vocabulary.len();
+        let mut vocab: Vec<String> = Vec::with_capacity(n);
```

Comment:
Any reason not to use a `Vec<(String, f64)>` directly?

Reply:
Nope seems like a good idea.